app/src/test/java/ch/hikemate/app/model/facilities/FacilitiesViewModelTest.kt

```diff
@@ -47,41 +48,80 @@ class FacilitiesViewModelTest {
     Dispatchers.resetMain()
   }
 
+  @OptIn(ExperimentalCoroutinesApi::class)
   @Test
-  fun testGetFacilities_onSuccess() {
-    val facilities = listOf(Facility(FacilityType.TOILETS, GeoPoint(46.51, 6.61)))
+  fun testGetFacilities_onSuccess() = runTest {
+    val testDispatcher = StandardTestDispatcher()
+    Dispatchers.setMain(testDispatcher)
 
-    `when`(mockFacilitiesRepository.getFacilities(any(), any(), any())).then {
-      val onSuccess = it.getArgument<(List<Facility>) -> Unit>(1)
-      onSuccess(facilities)
-    }
+    var onSuccessCalled = false
 
-    facilitiesViewModel.getFacilities(
-        bounds = testBounds,
-        onSuccess = { result -> assertEquals(facilities, result) },
-        onFailure = { fail("Should not be called") })
+    try {
+      val facilities = listOf(Facility(FacilityType.TOILETS, LatLong(46.51, 6.61)))
+
+      // Use suspending mock instead of callback-style
+      `when`(mockFacilitiesRepository.getFacilities(any(), any(), any())).then {
+        val onSuccess = it.getArgument<(List<Facility>) -> Unit>(1)
+        onSuccess(facilities)
+      }
+
+      facilitiesViewModel.getFacilities(
+          bounds = testBounds,
+          onSuccess = { result ->
+            assertEquals(facilities, result)
+            onSuccessCalled = true
+          },
+          onFailure = { fail("Should not be called") })
+
+      testDispatcher.scheduler.advanceUntilIdle()
+
+      assertTrue(onSuccessCalled)
+    } finally {
+      Dispatchers.resetMain()
+      assertTrue(onSuccessCalled)
+    }
   }
 
+  @OptIn(ExperimentalCoroutinesApi::class)
   @Test
   fun testGetFacilities_onFailure() {
     val exception = Exception("Test exception")
 
+    val testDispatcher = StandardTestDispatcher()
+    Dispatchers.setMain(testDispatcher)
+
     `when`(mockFacilitiesRepository.getFacilities(any(), any(), any())).then {
       val onFailure = it.getArgument<(Exception) -> Unit>(2)
       onFailure(exception)
     }
 
-    facilitiesViewModel.getFacilities(
-        bounds = testBounds,
-        onSuccess = { fail("Should not be called") },
-        onFailure = { ex -> assertEquals(exception, ex) })
+    var onFailureCalled = false
+
+    try {
+      facilitiesViewModel.getFacilities(
+          bounds = testBounds,
+          onSuccess = { fail("Should not be called") },
+          onFailure = { ex ->
+            assertEquals(exception, ex)
+            onFailureCalled = true
+          })
+
+      testDispatcher.scheduler.advanceUntilIdle()
+
+      assertTrue(onFailureCalled)
+    } finally {
+      Dispatchers.resetMain()
+      assertTrue(onFailureCalled)
+    }
```

Comment:
Same question as above

Reply:
Its `try...finally`. In order to not mess up other tests we need to clean up the Dispatchers. One way to do it is to add the Dispatcher setup to an `@Before` function and clean them up in an `@After` function, however since many tests don't use dispatchers that would be quite a bit of unnecessary boilerplate. I've decided that this method is more efficient. The finally clause always executes, even if there is an error within the test, so this way we guarantee that the dispatchers definitely get cleaned up before the next test starts.